packages/astro/src/runtime/client/dev-overlay/plugins/audit/a11y.ts

```diff
@@ -0,0 +1,610 @@
+			for (const attribute of requiredAttributes) {
+				if (!element.hasAttribute(attribute)) return true;
+			}
+
+			return false;
+		},
+	},
+	{
+		code: 'a11y-missing-content',
+		title: 'Missing content on element important for accessibility',
+		message: 'Headings and anchors must have content to be accessible.',
+		selector: a11y_required_content.join(','),
+		match(element) {
+			if (!element.textContent) return true;
+		},
+	},
+	{
+		code: 'a11y-no-redundant-roles',
+		title: 'HTML element has redundant ARIA roles',
+		message:
+			'Giving these elements an ARIA role that is already set by the browser has no effect and is redundant.',
+		selector: [...a11y_implicit_semantics.keys()].join(','),
+		match(element) {
+			const role = element.getAttribute('role');
+
+			if (element.localName === 'input') {
+				const type = element.getAttribute('type');
+				if (!type) return true;
+
+				const implicitRoleForType = input_type_to_implicit_role.get(type);
+				if (!implicitRoleForType) return true;
+
+				if (role === implicitRoleForType) return false;
+			}
+
+			// TODO: Handle menuitem and elements that inherit their role from their parent
+
+			const implicitRole = a11y_implicit_semantics.get(element.localName);
+			if (!implicitRole) return true;
+
+			if (role === implicitRole) return false;
+		},
+	},
+	{
+		code: 'a11y-no-interactive-element-to-noninteractive-role',
+		title: 'Non-interactive ARIA role used on interactive HTML element.',
+		message:
+			'Interactive HTML elements like `<a>` and `<button>` cannot use non-interactive roles like `heading`, `list`, `menu`, and `toolbar`.',
+		selector: `[role]:is(${interactiveElements.join(',')})`,
+		match(element) {
+			const role = element.getAttribute('role');
+			if (!role) return false;
+			if (!ariaRoles.has(role)) return false;
+
+			if (aria_non_interactive_roles.includes(role)) return true;
+		},
+	},
+	{
+		code: 'a11y-no-noninteractive-element-to-interactive-role',
+		title: 'Interactive ARIA role used on non-interactive HTML element.',
+		message: 'Interactive roles should not be used to convert a non-interactive element to an interactive element',
+		selector: `[role]:not(${interactiveElements.join(',')})`,
+		match(element) {
+			const role = element.getAttribute('role');
+			if (!role) return false;
+			if (!ariaRoles.has(role)) return false;
+			const exceptions = a11y_non_interactive_element_to_interactive_role_exceptions[element.localName as keyof typeof a11y_non_interactive_element_to_interactive_role_exceptions];
+			if (exceptions?.includes(role)) return false;
+
+			if (!aria_non_interactive_roles.includes(role)) return true;
+		},
+	},
+	{
+		code: 'a11y-no-noninteractive-tabindex',
+		title: 'Invalid `tabindex` on non-interactive element',
+		message: (element) => `${element.localName} elements should not have \`tabindex\` attribute`,
+		selector: '[tabindex]',
+		match(element) {
+			if (!interactiveElements.includes(element.localName)) return true;
+		},
+	},
```

Comment:
Maybe our code blocks need fixing 😠

Reply:
It's actually added for a11y reasons 😄 https://github.com/withastro/astro/issues/8001

Reply:
Oh yeah sorry, I meant "rule" as in the audit rule. It shouldn't generally consider scrollable elements to be interactive for this rule. Since the auditing is done client-side, there's no worries about the screen sizes and stuff.

Reply:
Oh yeah, sorry I think I was unclear, I made the audit "correct", it's still an accessibility issue for the code block to be tabbable when it's not scrollable! I believe we should fix the bug in our code blocks, not sure *how*, but we shouldn't hack around it because it is a valid issue.

Reply:
The rule is correct. To effectively address this issue, the optimal solution involves assessing the presence of horizontal scrollbars within `<pre>` elements on the client side. By dynamically adjusting the tabindex attribute based on this evaluation, we can ensure a seamless user experience. This client-side approach is essential due to the varying display characteristics across different screen sizes, which directly influence the visibility of horizontal scrollbars. Thus, by performing this assessment in the user's browser, we can accurately adapt to the specific rendering conditions and deliver consistent functionality regardless of the device or screen width.



```javascript
// Function to check if an element has a vertical scrollbar
function haveHorizontalScroll(element) {
	return element.scrollWidth > element.clientWidth;
}

// Function to manage the tabindex attribute of an element <pre> depending on whether it has a vertical scrollbar or not
function manageTabIndex(preElement) {
	if (haveHorizontalScroll(preElement)) {
		preElement.setAttribute("tabindex", "0");
		console.log("have horizontal scroll");
	} else {
		preElement.removeAttribute("tabindex");
		console.log("don't have horizontal scroll");
	}
}

// Function to observe all <pre> elements on the page
function observePreElements() {
	const preElements = document.querySelectorAll("pre.astro-code");
	preElements.forEach((preElement) => {
		const observer = new MutationObserver((mutationsList) => {
			for (let mutation of mutationsList) {
				if (mutation.type === "childList") {
					manageTabIndex(mutation.target);
				}
			}
		});

		const config = { childList: true, subtree: true };

		observer.observe(preElement, config);

		manageTabIndex(preElement);
	});
}
// Observe all <pre> elements on the page
observePreElements();
```